### PR TITLE
Fix Keyed Redis

### LINKED ISSUE #3873: Keyed Redis component doesn't work correctly
If I add 2 Redis server instances and try to use "keyed" services to talk to each, only the first connection is used for both. The 2nd connection doesn't get used.

```C#
var builder = DistributedApplication.CreateBuilder(args);

var basketCache = builder.AddRedis("basketcache");
var basketCache2 = builder.AddRedis("basketcache2");

var basketService = builder.AddProject("basketservice", @"..\BasketService\BasketService.csproj")
                           .WithReference(basketCache)
                           .WithReference(basketCache2);
```

And then in BasketService's Program.cs:

```C#
var builder = WebApplication.CreateBuilder(args);

builder.AddKeyedRedisClient("basketcache");
builder.AddKeyedRedisClient("basketcache2");

var app = builder.Build();

var m = app.Services.GetRequiredKeyedService<IConnectionMultiplexer>("basketcache");
var m2 = app.Services.GetRequiredKeyedService<IConnectionMultiplexer>("basketcache2");

Console.WriteLine(m.Configuration);
Console.WriteLine(m2.Configuration);
```

### PR DESCRIPTION
**Pull request by eerhardt**
When adding 2 Redis server instances and try to use "keyed" services to talk to each, only the first connection is used for both. The 2nd connection doesn't get used.

This is because we are using a singleton `OptionsFactory<ConfigurationOptions>` that only knows about the first StackExchangeRedisSettings registered, and always uses its ConnectionString.

The fix is to teach the factory about all the Settings objects being registered, and have it find the correct one given the "options name" when it is being created.

Fix #3873

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3885)

### CODE REVIEW

File: src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs [hunk 1]
```diff
@@ -189,17 +192,17 @@ private static ConfigurationOptions BindToConfiguration(ConfigurationOptions opt
...
             : base(setups, postConfigures, validations)
         {
-            _settings = settings;
+            _settingsPairs = settingsPairs;
         }
 
         protected override ConfigurationOptions CreateInstance(string name)
         {
-            var connectionString = _settings.ConnectionString;
+            var connectionString = _settingsPairs.FirstOrDefault(p => p.OptionsName == name)?.Settings.ConnectionString;
```
Comment (davidfowl) [inline src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:205]: What should the error be if there's no setting?

File: src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs [hunk 2]
```diff
@@ -80,17 +80,14 @@ public static class AspireRedisExtensions
...
         // see comments on ConfigurationOptionsFactory for why a factory is used here
-        builder.Services.TryAdd(ServiceDescriptor.Transient(typeof(IOptionsFactory<ConfigurationOptions>),
-            sp => new ConfigurationOptionsFactory(
-                settings,
-                sp.GetServices<IConfigureOptions<ConfigurationOptions>>(),
-                sp.GetServices<IPostConfigureOptions<ConfigurationOptions>>(),
-                sp.GetServices<IValidateOptions<ConfigurationOptions>>())));
-
-        string? optionsName = serviceKey is null ? null : connectionName;
+        builder.Services.AddTransient(sp => new RedisOptionsSettingsPair { OptionsName = optionsName, Settings = settings });
```
Comment (davidfowl) [inline src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:86]: This could use keyed di itself

Comment (eerhardt) [reply src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:86]: How? The IOptions system doesn't work with keyed DI.

File: src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs (hunk 1 above)
Comment (eerhardt) [reply src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:205]: We don't error here, but instead just create a "blank" ConfigurationOptions instance.

If there's no setting it basically means we have a bug, and not that the user did something wrong. I'm not sure how a user could get into that situation.

Comment (eerhardt) [reply src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:205]: > I guess someone could try serviceProvider.GetRequiredService<IOptionsMonitor<ConfigurationOptions>>().Get("doesNotExit").

Returning a blank ConfigurationOptions instance is consistent with other uses of named options that don't exist:

image

File: src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs (hunk 2 above)
Comment (Zombach) [reply src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:86]: Isn't that how it works?
```csharp
builder.Services.AddKeyedTransient<IOptions<Foo>>("key", (_, _) => Options.Create(new Fo0()));
builder.Services.AddKeyedTransient<IOptions<Foo>>("key2", (_, _) => Options.Create(new Fo0()));
```

Comment (eerhardt) [reply src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:86]: OK, I see what you mean. Just the `RedisOptionsSettingsPair` uses keyed DI, but not the OptionsFactory. I thought you meant the whole factory as well.

File: src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs (hunk 1 above)
Comment (eerhardt) [reply src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs]: I added a comment for this.

**Review by radical — APPROVED** [review]

**Pull request merged**